contributor-docs/adrs/adr-014-internal-modules.md

```diff
@@ -0,0 +1,62 @@
+# ADR 014: Internal Modules
+
+## Status
+
+Proposed
+
+## Context
+
+Currently all files live under the `src` directory. In the `npm` package for `@primer/react`, we specify the following export pattern:
+
+```json5
+{
+  "exports": {
+    // ...
+    "./lib-esm/*": {
+      "import": [
+        // ...
+      ],
+      "require": [
+        // ...
+      ]
+    }
+  }
+}
+```
+
+This pattern, along with our Rollup setup, opts-in files and folders under the `src` directory into the public API of the package. This means that certain parts of the codebase which aren't intended to be used outside of `@primer/react` are considered part of its public API and can be imported and used by consumers.
+
+## Decision
+
+Adopt a convention in Primer React where internal modules live in the `src/internal` folder. This folder which would include all components, hooks, and other modules which are not intended for usage outside of the project. Files within `src/internal` may be grouped by area, such as `src/internal/components`, `src/internal/hooks`, etc.
+
+In the `"exports"` field of our `npm` package, we can then add the following pattern:
+
+```json5
+{
+  "exports": {
+    // ...
+    "./lib-esm/internal/*": null
+  }
+}
+```
+
+This pattern would remove any files and folders within `src/internal` from the public API of the `npm` package. This pattern is inspired by [this section](https://nodejs.org/api/packages.html#package-entry-points) from Node.js, specifically this example:
+
+```json
+{
+  "name": "my-package",
+  "exports": {
+    ".": "./lib/index.js",
+    "./feature/*.js": "./feature/*.js",
+    "./feature/internal/*": null
+  }
+}
+```
+
+
+### Impact
+
+- Update the `"exports"` field in `package.json` to exclude `./lib-esm/internal` from usage
+- (In a major release) Move internal-only files into internal folder
```

Comment:
How would that affect the usages of these internal files? It makes total sense to do in a major release but I am wondering would we provide any alternatives or make exceptions to exclude some files if they are in used but intended to be internal? I hope it makes sense!

Reply:
Hey @broccolinisoup! 👋 

Great point, I'm glad you brought it up. If something is meant to be internal but it has usage in dotcom then we should definitely continue to export it from the bundle, in my opinion. I think at that point it stops being internal and is something that teams have come to rely on (albeit accidentally) and clearly has value for them and for us to support.

Curious what you think 👀

Reply:
Yeah, I agree!